From 1f426693c14c79e65ad8e62ef6bb4b07d6484040 Mon Sep 17 00:00:00 2001
Message-Id: <1f426693c14c79e65ad8e62ef6bb4b07d6484040.1440106296.git.chang-joon.lee@intel.com>
From: Michel Thierry <michel.thierry@intel.com>
Date: Fri, 10 Jul 2015 16:27:15 +0100
Subject: [PATCH 01/11] REVERTME [VPG]: drm/i915: prevent out of range px in PX
 macros (take 2)

In upstream we tried to fix this in the following commit:

commit fdc454c1484a20e1345cf4e4d7a9feaee814147f
Author: Michel Thierry <michel.thierry <at> intel.com>
Date:   Tue Mar 24 15:46:19 2015 +0000
    drm/i915: Prevent out of range pt in gen6_for_each_pde

but the static analyzer still complains that, just before we break due
to "iter < I915_PDES", we do "pt = (pd)->page_table[iter]" with an
iter value that is bigger than I915_PDES. Of course, this isn't really
a problem since no one uses pt outside the macro. Still, every single
new usage of the macro will create a new issue for us to mark as a
false positive.

REVERTME: Because this tree is no longer updated and is missing a lot of
changes for ppgtt. This patch is a port, and if a forklift ever happen
again, it is better to just drop this patch.

Change-Id: I69217979fcc956dc188a190f2e3b6c8f228bb730
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-40946
Signed-off-by: Michel Thierry <michel.thierry@intel.com>

## Don't forget to add a JIRA URL to the Tracked-On metadata!
## https://wiki.ith.intel.com/display/CACTUS/Patch+requirements+check
## url should be in the form:
## Tracked-On: https://jira01.devtools.intel.com/browse/<PROJ>-<NUMB>
## Several Tracked-On lines are allowed
## you can fully prefill your Tracked-On commits using environment variable:
## export CURRENT_TRACKED_ON=https://jira01.devtools.intel.com/browse/<PROJ>-<NUMB>
##
---
 drivers/gpu/drm/i915/i915_gem_gtt.h | 31 +++++++++++++++++++------------
 1 file changed, 19 insertions(+), 12 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_gem_gtt.h b/drivers/gpu/drm/i915/i915_gem_gtt.h
index 49e67cd..eef9744 100644
--- a/drivers/gpu/drm/i915/i915_gem_gtt.h
+++ b/drivers/gpu/drm/i915/i915_gem_gtt.h
@@ -312,17 +312,20 @@ struct i915_hw_ppgtt {
  * XXX: temp is not actually needed, but it saves doing the ALIGN operation.
  */
 #define gen6_for_each_pde(pt, pd, start, length, temp, iter) \
-	for (iter = gen6_pde_index(start), pt = (pd)->page_tables[iter]; \
+	for (iter = gen6_pde_index(start); \
+	     pt = (length > 0 && iter < GEN6_PPGTT_PD_ENTRIES) ? \
+	     (pd)->page_tables[iter] : NULL, \
 	     length > 0 && iter < GEN6_PPGTT_PD_ENTRIES; \
-	     pt = (pd)->page_tables[++iter], \
+	     iter++, \
 	     temp = ALIGN(start+1, 1 << GEN6_PDE_SHIFT) - start, \
 	     temp = min_t(unsigned, temp, length), \
 	     start += temp, length -= temp)
 
 #define gen6_for_all_pdes(pt, ppgtt, iter)  \
-	for (iter = 0, pt = ppgtt->pd.page_tables[iter];			\
-	     iter < gen6_pde_index(ppgtt->base.total);			\
-	     pt =  ppgtt->pd.page_tables[++iter])
+	for (iter = 0;		\
+	     pt = (iter < GEN6_PPGTT_PD_ENTRIES) ? ppgtt->pd.page_tables[iter] : NULL, \
+	     iter < GEN6_PPGTT_PD_ENTRIES;	\
+	     iter++)
 
 static inline uint32_t i915_pte_index(uint64_t address, uint32_t pde_shift)
 {
@@ -373,17 +376,21 @@ static inline uint32_t gen6_pde_index(uint32_t addr)
 }
 
 #define gen8_for_each_pde(pt, pd, start, length, temp, iter)		\
-	for (iter = gen8_pde_index(start), pt = (pd)->page_tables[iter]; \
-	     length > 0 && iter < GEN8_PDES_PER_PAGE;			\
-	     pt = (pd)->page_tables[++iter],				\
+	for (iter = gen8_pde_index(start); \
+	     pt = (length > 0 && iter < GEN8_PDES_PER_PAGE) ? \
+	     (pd)->page_tables[iter] : NULL, \
+	     length > 0 && iter < GEN8_PDES_PER_PAGE;	\
+	     iter++,				\
 	     temp = ALIGN(start+1, 1 << GEN8_PDE_SHIFT) - start,	\
 	     temp = min(temp, length),					\
 	     start += temp, length -= temp)
 
-#define gen8_for_each_pdpe(pd, pdp, start, length, temp, iter)		\
-	for (iter = gen8_pdpe_index(start), pd = (pdp)->page_directory[iter];	\
-	     length > 0 && iter < GEN8_LEGACY_PDPES;			\
-	     pd = (pdp)->page_directory[++iter],				\
+#define gen8_for_each_pdpe(pd, pdp, start, length, temp, iter)	\
+	for (iter = gen8_pdpe_index(start); \
+	     pd = (length > 0 && iter < GEN8_LEGACY_PDPES) ? \
+	     (pdp)->page_directory[iter] : NULL, \
+	     length > 0 && iter < GEN8_LEGACY_PDPES;	\
+	     iter++,				\
 	     temp = ALIGN(start+1, 1 << GEN8_PDPE_SHIFT) - start,	\
 	     temp = min(temp, length),					\
 	     start += temp, length -= temp)
-- 
1.9.1

